File: solutions/python/19.py
def reveal_santa_route(routes: list[list[str]]) -> list[str]:
    mapping = {origin: dest for origin, dest in routes}

    dests = {dest for _, dest in routes}
    start = next(origin for origin, _ in routes if origin not in dests)

    result = [start]
    while start in mapping:
        start = mapping[start]
        result.append(start)

    return result


reveal1 = reveal_santa_route([
    ["MEX", "CAN"],
    ["UK", "GER"],
    ["CAN", "UK"],
])
# → ['MEX', 'CAN', 'UK', 'GER']
print(reveal1)

reveal2 = reveal_santa_route([
    ["USA", "BRA"],
    ["JPN", "PHL"],
    ["BRA", "UAE"],
    ["UAE", "JPN"],
    ["CMX", "HKN"],
])
# → ['USA', 'BRA', 'UAE', 'JPN', 'PHL']
print(reveal2)

reveal3 = reveal_santa_route([
    ["STA", "HYD"],
    ["ESP", "CHN"],
])
# → ['STA', 'HYD']
print(reveal3)


"""
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 240
  ops: 3109
"""

File: solutions/ts/19.ts
function revealSantaRoute(routes: string[][]): string[] {
  const validRoutes = [...routes[0]];
  let pendingRoutes = routes.slice(1);
  let finished = false;
  while (!finished) {
    const nextRouteIndex = pendingRoutes.findIndex(
      (route) => route[0] === validRoutes.at(-1)
    );
    if (nextRouteIndex > -1) {
      validRoutes.push(pendingRoutes[nextRouteIndex][1]);
      pendingRoutes.splice(nextRouteIndex, 1);
    } else {
      finished = true;
    }
  }
  return validRoutes;
}

const reveal1 = revealSantaRoute([
  ["MEX", "CAN"],
  ["UK", "GER"],
  ["CAN", "UK"],
]);
// → ['MEX', 'CAN', 'UK', 'GER']
console.log(reveal1);

const reveal2 = revealSantaRoute([
  ["USA", "BRA"],
  ["JPN", "PHL"],
  ["BRA", "UAE"],
  ["UAE", "JPN"],
  ["CMX", "HKN"],
]);
// → ['USA', 'BRA', 'UAE', 'JPN', 'PHL']
console.log(reveal2);

const reveal3 = revealSantaRoute([
  ["STA", "HYD"],
  ["ESP", "CHN"],
]);
// → ['STA', 'HYD']
console.log(reveal3);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 340
  ops: 3887
*/
